File: apps/backend/src/executors/dbExecutor.ts
import { Client } from 'pg';
import type { ExecutorResult } from './restExecutor.js';

export interface DbExecutorInput {
  connectionString: string;
  query:            string;
  params?:          unknown[];
}

// Any keyword that mutates schema or data.
// Checked before the query runs as a fast-fail; BEGIN READ ONLY is the
// authoritative enforcement layer that catches anything this misses.
const WRITE_RE = /\b(INSERT|UPDATE|DELETE|DROP|ALTER|TRUNCATE|CREATE|REPLACE|GRANT|REVOKE|MERGE|COPY)\b/i;

export async function dbExecutor(input: DbExecutorInput): Promise<ExecutorResult> {
  const { connectionString, query, params = [] } = input;

  if (!connectionString) {
    return { success: false, error: 'No connection string resolved for this resource — check secret_ref and the corresponding env variable' };
  }

  if (WRITE_RE.test(query)) {
    return {
      success: false,
      error:   'Write operations are not permitted. Only SELECT queries are allowed through the execute API.',
    };
  }

  const client = new Client({ connectionString });

  try {
    await client.connect();

    // Defence layer 2: run inside a read-only transaction so the database itself
    // enforces the constraint even if the regex above is somehow bypassed.
    await client.query('BEGIN READ ONLY');
    const result = await client.query(query, params);
    await client.query('COMMIT');

    return { success: true, data: result.rows };
  } catch (err) {
    // Roll back cleanly on query errors (no-op if the transaction never opened)
    try { await client.query('ROLLBACK'); } catch { /* ignore */ }
    return { success: false, error: (err as Error).message };
  } finally {
    try { await client.end(); } catch { /* ignore */ }
  }
}

File: apps/backend-py/app/executors/db_executor.py
import re
from typing import Any

import asyncpg

WRITE_RE = re.compile(
    r"\b(INSERT|UPDATE|DELETE|DROP|ALTER|TRUNCATE|CREATE|REPLACE|GRANT|REVOKE|MERGE|COPY)\b",
    re.IGNORECASE,
)


async def db_executor(
    connection_string: str,
    query: str,
    params: list[Any] | None = None,
) -> dict[str, Any]:
    if not connection_string:
        return {
            "success": False,
            "error": "No connection string resolved for this resource — check secret_ref and the corresponding env variable",
        }

    if WRITE_RE.search(query):
        return {
            "success": False,
            "error": "Write operations are not permitted. Only SELECT queries are allowed through the execute API.",
        }

    conn: asyncpg.Connection | None = None
    try:
        conn = await asyncpg.connect(dsn=connection_string)
        await conn.execute("BEGIN READ ONLY")
        rows = await conn.fetch(query, *(params or []))
        await conn.execute("COMMIT")
        return {"success": True, "data": [dict(r) for r in rows]}
    except Exception as exc:
        if conn:
            try:
                await conn.execute("ROLLBACK")
            except Exception:
                pass
        return {"success": False, "error": str(exc)}
    finally:
        if conn:
            try:
                await conn.close()
            except Exception:
                pass
